File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
public:
        int dfs(int r, int c, int current, vector<vector<int>>& grid) {
            if(r < 0 || c < 0 || r >= grid.size() || c >= grid[r].size()) {
                return current;
            }
            if(grid[r][c] == 0) {
                return current;
            }
            current++;
            grid[r][c] = 0; //mark as visited
            return std::max({
                dfs(r + 1, c, current, grid),
                dfs(r - 1, c, current, grid),
                dfs(r, c + 1, current, grid),
                dfs(r, c - 1, current, grid)
        });
        }

    int maxAreaOfIsland(vector<vector<int>>& grid) {

        int m = 0;
        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[i].size(); j++) {
                int current = dfs(i, j, 0, grid);
                if(current > m) {
                    m = current;
                }
            }
        }
        return m;

    }
};

File: Data Structures & Algorithms/surrounded-regions/submission-1.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        # Run BFS from each of the O cells, check if any of them reach the border and if so stay the same, else recursively swap  
        global_visited = set()
        directions = [
            (0, 1),
            (0, -1),
            (1, 0),
            (-1, 0)
        ]

        def bfs(r, c, board):
            #Start with implementing basic dfs
            visited = set()
            queue = collections.deque([(r, c)])
            visited.add((r, c))
            global_visited.add((r, c))
            touched_border = False
            while queue:
                current = queue.popleft()
                if current[0] == 0 or current[0] == len(board) - 1 or current[1] == 0 or current[1] == len(board[0]) - 1:
                    touched_border = True
                for dir in directions:
                    row = current[0] + dir[0]
                    col = current[1] + dir[1]
                    if 0 <= row < len(board) and 0 <= col < len(board[0]):
                        if board[row][col] == 'O' and (row, col) not in visited:
                            queue.append((row, col))
                            visited.add((row, col))
                            global_visited.add((row, col))
            if(not touched_border):
                for cell in visited:
                    board[cell[0]][cell[1]] = 'X'

        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j] == 'O' and (i, j) not in global_visited:
                    bfs(i, j, board)